core/src/mast/merger/mod.rs

```diff
@@ -0,0 +1,409 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+
+use miden_crypto::hash::{blake::Blake3Digest, rpo::RpoDigest};
+
+use crate::mast::{
+    DecoratorId, EqHash, MastForest, MastForestError, MastNode, MastNodeId, MultiMastForestNodeIter,
+};
+
+#[cfg(test)]
+mod tests;
+
+/// A type that allows merging [`MastForest`]s.
+///
+/// This functionality is exposed via [`MastForest::merge`]. See its documentation for more details.
+pub(crate) struct MastForestMerger {
+    mast_forest: MastForest,
+    // Internal indices needed for efficient duplicate checking and EqHash computation.
+    node_id_by_hash: BTreeMap<RpoDigest, Vec<(EqHash, MastNodeId)>>,
+    hash_by_node_id: BTreeMap<MastNodeId, EqHash>,
+    decorators_by_hash: BTreeMap<Blake3Digest<32>, DecoratorId>,
+    decorator_id_mappings: Vec<DecoratorIdMap>,
+    node_id_mappings: Vec<MastForestNodeIdMap>,
+}
+
+impl MastForestMerger {
+    /// Creates a new merger which creates a new internal, empty forest into which other
+    /// [`MastForest`]s are merged.
+    pub(crate) fn merge<'forest>(
+        forests: impl IntoIterator<Item = &'forest MastForest>,
+    ) -> Result<(MastForest, Vec<MastForestRootMap>), MastForestError> {
+        let forests = forests.into_iter().collect::<Vec<_>>();
+        let decorator_id_mappings = Vec::with_capacity(forests.len());
+        let node_id_mappings = vec![MastForestNodeIdMap::new(); forests.len()];
+
+        let mut merger = Self {
+            node_id_by_hash: BTreeMap::new(),
+            hash_by_node_id: BTreeMap::new(),
+            decorators_by_hash: BTreeMap::new(),
+            mast_forest: MastForest::new(),
+            decorator_id_mappings,
+            node_id_mappings,
+        };
+
+        merger.merge_inner(forests.clone())?;
+
+        let Self { mast_forest, node_id_mappings, .. } = merger;
+
+        let mut root_maps = Vec::new();
+        for (forest_idx, mapping) in node_id_mappings.into_iter().enumerate() {
+            let forest = forests[forest_idx];
+            root_maps.push(MastForestRootMap::from_node_id_map(mapping, &forest.roots));
+        }
+
+        Ok((mast_forest, root_maps))
+    }
+
+    fn merge_inner(&mut self, forests: Vec<&MastForest>) -> Result<(), MastForestError> {
```

Comment:
Could we add doc comments to this method (especially covering the `ExternalNodeReplacement` case)?